src/renderers/webgl/WebGLProgram.js

```diff
@@ -422,6 +424,7 @@ function WebGLProgram( renderer, cacheKey, parameters, bindingStates ) {
 
 		prefixFragment = [
 
+			versionString,
```

Comment:
`versionString` needs also to be added to `ShaderMaterial`. Right now, it's not possible to write GLSL 3.0 shader with `ShaderMaterial` because of #19889.

Reply:
Okay, I'm trying to understand what you are suggesting should be done. Currently starting at line 678 , there's some checks done for GLSL 3.0 conversions for non-RawShaderMaterials. Within my testing, I did notice that standard ShaderMaterials were transformed to GLSL 3.0 properly. However do you mean that you cannot use GLSL 1.0 code within a WebGL2 context? That would be definitely be the case since it automatically converts it over to GLSL 3.0.

So should I modify those lines to:

1) check for the `glslVersion` property to perform the conversion conditionally only if THREE.GLSL3 is used.  There could be side effects, but this seems like the best approach long term.

2) remove the `#version 300 es` from the prefixString and use `versionString` only.

3) keep the conversion code as is (minus the version string).

Reply:
> However do you mean that you cannot use GLSL 1.0 code within a WebGL2 context?

No, I mean the following fiddle should work but it doesn't starting with `r119`: https://jsfiddle.net/1q3gs0a6/2/

We did not realize that merging #19889 has side effects. The previous code ( `r118`) ensured that the above code works (see https://jsfiddle.net/d8v1j5x7/).

Hence, the engine needs a replacement. If `ShaderMaterial` is used and if `glslVersion` is set to `THREE.GLSL3`, the following code shall not be defined:
```js
out highp vec4 pc_fragColor;
#define gl_FragColor pc_fragColor
``` 
I _believe_ this should fix the fiddle.

Reply:
Okay, so in my list of changes, you would want 1), 2) and 3) applied. This fix this fiddle, the `#version 300 es` must be removed from the top because it will not compile with this statement if it is not the first line (even if the program already begins with `#version 300 es` as the first line). With regards to preventing the additional conversion with the 'glslVersion' check, that will work.

Reply:
One unintended side effect of changing this code would be breaking the internal shader chunks that rely on GLSL 3.0 compatibility.

I've chosen to change the conditional logic to test for `glslVersion == null` and only apply the conversion is this case. It might seem strange at first, but if you think about it, having the GLSL version unspecified would denote that the code is flexible enough to be treated as any possible version. 

The prepending bugs should be solved moving forward so long as:

1) a user supplies the `glslVersion` property without adding `#version XXX` to their programs and they don't need automated conversion.

2) a user doesn't specify a `glslVersion`
2a) if the user uses a RawShaderMaterial, they cannot reliably use the `defines` feature and will have to supply `#version XXX` as necessary.
2b) if the user uses a ShaderMaterial, the user should not ever specify the `#version XXX` in the program and allow for automated conversion as necessary.

Reply:
Sorry, but I have the feeling this is getting too complicated. I don't think `glslVersion` should be `null`. The default should be `THREE.GLSL1`.

Using `ShaderMaterial` and `RawShaderMaterial` means the user is responsible for the shader code. Hence, he has to specify whether it is GLSL 1.0 or GLSL 3.0 code. 

Requesting the user to remove the GLSL version strings from his shader code seems appropriate according to what @mrdoob says here: https://github.com/mrdoob/three.js/issues/19959#issuecomment-667193339

Reply:
@Mugen87 okay, so how do you propose fixing the internal shader chunks? Do all of those broken ShaderMaterials have to specify `glslVersion=THREE.GLSL3`?

Reply:
Using the `glslVersion = null` as the default would keep compatibility with any currently working code. When it is unspecified, no functionality changes. Your Fiddle example is broken regardless because there's no ideal solution for supporting `#version 300 es` embeded in a ShaderMaterial anyways.

Reply:
>  Do all of those broken ShaderMaterials have to specify glslVersion=THREE.GLSL3?

If I understood @mrdoob correctly, then yes. It's the replacement for defining the version in the shader code.

> . Your Fiddle example is broken regardless because there's no ideal solution for supporting #version 300 es embeded in a ShaderMaterial anyways.

The previous solution which was #19889 removed with `r118` will probably not be added back since @mrdoob wants to avoid the related complexity in `WebGLProgram`. Hence a migration task will be required. Asking users to define `Material.glslVersion` seems like an acceptable solution.

Reply:
> Any users with RawShaderMaterials out in the wild will have to change their code to support this new property. 

I'm not sure I understand this. It only affects users who explicitly developed GLSL 3.0 code for WebGL 2.

Reply:
Graceful migration is an important consideration here. THREE.js offers some backwards compatibility when making changes at least for some time in my experience with it. All this debate is focused on changing one line of code here, which will cause other consequences. I think @mrdoob should review and comment since r119 change was also based on his feedback.

Reply:
@Mugen87 yes that is correct, it affects compatibility of RawShaderMaterials for WebGL2 and GLSL 3.0. Since neither of those are relatively new, there would be compatibility affects for users. I know because I'm one of those people affected already since I have other code that I have to change based on this change.

Reply:
I vote to revert #19889 for now and then think of a clean solution which is backwards compatible. Detecting the version string of the shader code like before seems absolutely fine, IMO. 

I don't like breaking changes if they are avoidable which is absolutely possible in this case.

Reply:
Again, this change offer NO breaking change. Changing `glslVersion` to default to non-null and making an explicit check for 'glslVersion` would be breaking change. I hope that is clear from all the previous comments. Code has been tested in a project that uses various types of shaders with no side effects.